device/cuda/src/details/sparse_ccl.cu

```diff
@@ -0,0 +1,343 @@
+            outi = 0;
+        }
+
+        /*
+         * These arrays are the meat of the pudding of this algorithm, and we
+         * will constantly be writing and reading from them which is why we
+         * declare them to be in the fast shared memory. Note that this places a
+         * limit on the maximum activations per module, as the amount of shared
+         * memory is limited. These could always be moved to global memory, but
+         * the algorithm would be decidedly slower in that case.
+         */
+        __shared__ unsigned int f[MAX_ACTIVATIONS_PER_MODULE], gf[MAX_ACTIVATIONS_PER_MODULE];
+
+        /*
+         * Here we initialize a few variables. This loop pattern is commonly
+         * found throughout this code. We iterate over the number of activations
+         * each thread must process. Sadly, the CUDA limit is 1024 threads per
+         * block and we cannot guarantee that there will be fewer than 1024
+         * activations in a module. So each thread must be able to do more than
+         * one.
+         */
+        for (unsigned int tst = 0; tst * blockDim.x + threadIdx.x < size; ++tst) {
+            /*
+             * This line calculates the activation or cell ID from the thread ID
+             * and the value in the outer loop.
+             */
+            unsigned int tid = tst * blockDim.x + threadIdx.x;
+
+            /*
+             * At the start, the values of f and gf should be equal to the ID of
+             * the cell.
+             */
+            f[tid] = tid;
+            gf[tid] = tid;
+
+            /*
+             * The number of adjacent cells for each cell must start at zero, to
+             * avoid uninitialized memory. adjv does not need to be zeroed, as
+             * we will only access those values if adjc indicates that the value
+             * is set.
+             */
+            adjc[tst] = 0;
+        }
+
+        /*
+         * After this is all done, we synchronise the block. I am not absolutely
+         * certain that this is necessary here, but the overhead is not that big
+         * and we might as well be safe rather than sorry.
+         */
+        __syncthreads();
+
+        /*
+         * This loop initializes the adjacency cache, which essentially
+         * translates the sparse CCL problem into a graph CCL problem which we
+         * can tackle with well-studied algorithms.
+         */
+        for (unsigned int tst = 0; tst * blockDim.x + threadIdx.x < size; ++tst) {
+            unsigned int tid = tst * blockDim.x + threadIdx.x;
+
+            unsigned int tc0 = cells[tid].channel0;
+            unsigned int tc1 = cells[tid].channel1;
+
+            /*
+             * First, we traverse the cells backwards, starting from the current
+             * cell and working back to the first, collecting adjacent cells
+             * along the way.
+             */
+            for (unsigned int j = tid - 1; j < tid; --j) {
```

Comment:
`j >= 0`? Guessing this works because `unsigned int` wraps around, in fact becoming greater than `tid`.

Reply:
`i >= 0` on decrementing unsigned integer loops is an anti-pattern and will keep looping forever. This is how I've always been taught to do loops for unsigned integers.